elements/lisk-chain/src/data_access/transaction_interface_adapter.ts

```diff
@@ -24,25 +26,52 @@ export class TransactionInterfaceAdapter {
 	public constructor(registeredTransactions: RegisteredTransactions = {}) {
 		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
 		this._transactionClassMap = new Map();
+		codec.addSchema(BaseTransaction.BASE_SCHEMA);
 		Object.keys(registeredTransactions).forEach(transactionType => {
+			const transaction = registeredTransactions[transactionType];
 			this._transactionClassMap.set(
 				Number(transactionType),
-				registeredTransactions[transactionType],
+				transaction,
 			);
+			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+			const { ASSET_SCHEMA } = this._transactionClassMap.get(Number(transactionType));
+			codec.addSchema(ASSET_SCHEMA);
 		});
 	}
 
-	public fromJSON(rawTx: TransactionJSON): BaseTransaction {
+	// First encode message asset and then encode base message
+	public encode(message: BaseTransaction): Buffer {
 		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
-		const TransactionClass = this._transactionClassMap.get(rawTx.type);
+		const TransactionClass = this._transactionClassMap.get(message.type);
 
 		if (!TransactionClass) {
 			throw new Error('Transaction type not found.');
 		}
 
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+		const binaryAsset = codec.encode(TransactionClass.ASSET_SCHEMA, message.asset as unknown as GenericObject);
+		const binaryAssetWithMessage = { ...message, asset: binaryAsset };
+
+		const binaryMessage = codec.encode(BaseTransaction.BASE_SCHEMA, binaryAssetWithMessage as unknown as GenericObject);
+
+		return binaryMessage;
+	}
+
+	// First decode base message and then decode asset
```

Comment:
Why is this done in two steps instead of doing it in one `decode()` call? I think it might be faster in general to call decode only once if possible

Reply:
you mean, create the whole schema and decode? I followed the LIP, according to LIP we need to decode root message first and then decode asset https://github.com/LiskHQ/lips/blob/master/proposals/lip-0028.md#deserialization